service/history/timer_queue_active_task_executor.go

```diff
@@ -230,19 +230,13 @@ func (t *timerQueueActiveTaskExecutor) executeActivityTimeoutTask(
 	if mutableState == nil || !mutableState.IsWorkflowExecutionRunning() {
 		return nil
 	}
-	ai, ok := mutableState.GetActivityInfo(task.EventID)
+	_, ok := mutableState.GetActivityInfo(task.EventID)
 
 	if !ok {
 		// if activity is not found, the timer is invalid.
 		return nil
 	}
 
-	if task.Stamp != ai.Stamp {
-		// if this task is invalid - we want to generate new activity timeout task
-		// it will be done in closeTransactionPrepareTasks call
-		return weContext.UpdateWorkflowExecutionAsActive(ctx, t.shardContext)
-	}
```

Comment:
Ah this makes sense. So basically when you `LoadAndSortActivityTimers()` you'll generate only relevant timer tasks.

Maybe put a comment in this function explaining that there's no need to check the stamp here since we generate the timer sequence on the fly and it is always valid given the current activity state.

Reply:
add a comment